packages/xstate-wallet/src/workflows/ethereum-enable.ts

```diff
@@ -0,0 +1,128 @@
+// A workflow to guide the user through enabling window.ethereum
+
+import {
+  StateSchema,
+  State,
+  Action,
+  MachineConfig,
+  Machine,
+  StateMachine,
+  ServiceConfig,
+  assign
+} from 'xstate';
+import {sendDisplayMessage, MessagingServiceInterface} from '../messaging';
+import {Store} from '../store';
+
+interface UserApproves {
+  type: 'USER_APPROVES_ENABLE';
+}
+interface UserRejects {
+  type: 'USER_REJECTS_ENABLE';
+}
+export type WorkflowEvent = UserApproves | UserRejects;
+
+export interface WorkflowContext {
+  requestId: number;
+  enabledAddress?: string;
+}
+
+export interface WorkflowServices extends Record<string, ServiceConfig<WorkflowContext>> {
+  enableEthereum: () => Promise<string>;
+}
+export interface WorkflowStateSchema extends StateSchema<WorkflowContext> {
+  states: {
+    explainToUser: {};
+    enabling: {};
+    done: {};
+    failure: {};
+  };
+}
+export type WorkflowState = State<WorkflowContext, WorkflowEvent, WorkflowStateSchema, any>;
+
+export interface WorkflowActions {
+  hideUi: Action<WorkflowContext, any>;
+  displayUi: Action<WorkflowContext, any>; // calls directly to messaging service
+  sendResponse: Action<WorkflowContext, any>; // calls directly to messaging service
+}
+export type StateValue = keyof WorkflowStateSchema['states'];
+
+const generateConfig = (
+  actions: WorkflowActions
+): MachineConfig<WorkflowContext, WorkflowStateSchema, WorkflowEvent> => ({
+  id: 'enable-ethereum',
+  initial: 'explainToUser',
+  states: {
+    explainToUser: {
+      entry: [actions.displayUi],
+      on: {
+        USER_APPROVES_ENABLE: {target: 'enabling'},
+        USER_REJECTS_ENABLE: {target: 'failure'}
+      }
+    },
+    enabling: {
+      invoke: {
+        src: 'enableEthereum',
+        onDone: {target: 'done', actions: assign({enabledAddress: (context, event) => event.data})},
+        onError: 'failure'
+      }
+    }, // invoke ethereum enable
+    done: {type: 'final', entry: [actions.hideUi, actions.sendResponse]},
+    failure: {type: 'final'}
```

Comment:
<img width="833" alt="Screenshot 2020-03-12 at 15 08 31" src="https://user-images.githubusercontent.com/1833419/76535930-8db9c480-6473-11ea-9c44-761a8cc93ef7.png">

This seems unfinished? Maybe we should `actions.hideUI` and signal to the app to go back to a state where it can try again.

Reply:
I've decided to handle this in a separate PR, as it'll include the necessary changes to send an error response back over the json-rpc connection.